runtime/compiler/env/PersistentAllocator.cpp

```diff
@@ -68,78 +65,133 @@ PersistentAllocator::allocate(size_t size, void *hint)
    return alloc;
    }
 
+PersistentAllocator::Block *
+PersistentAllocator::allocateFromVariableSizeListLocked(size_t allocSize)
+   {
+   Block * block = NULL;
+   Block * prev = NULL;
+   for (block = _freeBlocks[0]; block && block->_size < allocSize; prev = block, block = prev->next())
+      {}
+
+   if (block)
+      {
+      if (prev)
+         prev->_next = block->next();
+      else
+         _freeBlocks[0] = block->next();
+
+      block->_next = NULL;
+      }
+   return block;
+   }
+
 void *
-PersistentAllocator::allocateLocked(size_t requestedSize)
+PersistentAllocator::allocateInternal(size_t requestedSize)
    {
    TR_ASSERT( sizeof(Block) == mem_round( sizeof(Block) ),"Persistent block size will prevent us from properly aligning allocations.");
    size_t const dataSize = mem_round(requestedSize);
    size_t const allocSize = sizeof(Block) + dataSize;
 
-   TR::AllocatedMemoryMeter::update_allocated(allocSize, persistentAlloc);
+   if (TR::AllocatedMemoryMeter::_enabled & persistentAlloc)
+      {
+      // Use _smallBlockListsMonitor to protect TR::AllocatedMemoryMeter::update_allocated
+      // because accessing the variable-size-block list protected by ::memoryAllocMonitor
+      // takes longer and we may be penalizing access to fixed size block which should be very fast
+      j9thread_monitor_enter(_smallBlockListsMonitor);
```

Comment:
Persistent memory is created very early when TR monitors don't even exist. This is why that ugliness
```
      if (::memoryAllocMonitor)
         ::memoryAllocMonitor->enter();
```
exists. In here I decided to use VM constructs for the mutexes, rather than repeating the pattern above.
Due to these complications `OMR::Critical Section(_smallBlockListsMonitor);` will not work.

Reply:
Ah....that's unfortunate :/